### DISCUSSION TITLE
SWIM cluster for automatic clustering

### DISCUSSION THREAD

**Discussion by zbalkan** [rqlite/rqlite#1414] [Ideas]
The [SWIM protocol](https://www.cs.cornell.edu/projects/Quicksilver/public_pdfs/SWIM.pdf) is a membership protocol for clusters based on gossip. It allows the cluster manage the membership internally and automatically.

This may be another option to have another simple dynamic clustering strategy. I will leave here a few good references. The good thing for rqlite is the [memberlist by HashiCorp](https://github.com/hashicorp/memberlist), which is a robust implementation written in Go.

The pro here is dynamic membership allowing fault tolerance. In a scale out expansion, new nodes can join the cluster as soon as they are ready. And also, they can be removed from the membership when they are down. The member list must be shared with Raft member list for a clear and robust cluster.

References:

- https://www.bartoszsypytkowski.com/make-your-cluster-swim/
- https://blog.kevingomez.fr/2019/01/29/clusters-and-membership-discovering-the-swim-protocol/
- https://www.hashicorp.com/resources/everybody-talks-gossip-serf-memberlist-raft-swim-hashicorp-consul

**Comment by zbalkan** [comment]
I would like to mention a .NET project which combines RAFT and SWIM as a library to provide this capability: https://github.com/zarusz/SlimCluster

**Comment by otoolep** [comment]
Thanks for the reference -- I'll look into it.
